lib/util/core/src/format.rs

```diff
@@ -54,3 +56,18 @@ pub fn truncate_at_code_point(
 			.collect(),
 	)
 }
+
+pub fn item_list<'a, I: Iterator<Item = impl AsRef<str>>>(mut iter: I) -> String {
+	let mut s = String::new();
+
+	if let Some(item) = iter.next() {
+		s.push_str(item.as_ref());
+	}
+
+	while let Some(item) = iter.next() {
+		s.push_str(item.as_ref());
+		s.push_str(", ");
```

Comment:
It seems like this will add a trailing comma at the end if there is more than one element in the iterator. If you don't want the trailing comma, then I think this should be more concise?

```rust
pub fn item_list<'a, I: Iterator<Item = impl AsRef<str>>>(iter: I) -> String {
    iter.map(|item| item.as_ref().to_string())
        .collect::<Vec<_>>()
        .join(", ")
}
```

Reply:
Its more concise but allocates a vector. The problem with my code is i put the comma push after the item, it should be before since i added the `if let Some` before the loop to prevent a tailing/leading comma

Reply:
Hmm, I think your code allocates on the heap as well, if that's what you mean. [`String` uses a `vec` under the hood](https://doc.rust-lang.org/src/alloc/string.rs.html#365). I think they should be relatively identical, mine might be slightly more cause I'm mapping to a string, but I imagine it shouldn't be too different on any benchmarks.

Reply:
Did a check, and looks like mine does have more of a performance cost. Neat to check out.

![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/XiP77c3NM4Q27xw9sqIb/6c550778-82a5-4fd7-84e3-ca30be4f37d3.png)

Reply:
Your code allocates a string for each item, then a vector for all the strings, then a string to combine the vector items. Might be worth removing the `to_string` and testing again, but its still 1 more allocation with the vector

Reply:
Im guessing mine could be improved also with something like `String::with_capacity(iter.clone().fold(0, |item| item.len()))` to prevent extending the underlying vector with each `push_str` but thats excessive